calico.go

```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
+	if orchestrator == "k8s" {
+		wep, err = calicoClient.WorkloadEndpoints().Get(ep)
 		if err != nil {
-			return err
+			if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
+
+				// We can talk to the datastore but WEP doesn't exist in there,
+				// but we still want to go ahead with the clean up. So, log a warning and clean up.
+				logger.WithField("WorkloadEndpoint", ep).Warningf("Error getting workloadendpoint: %v", err)
+			} else {
+				// Could not connect to datastore (connection refused, unauthorized, etc.)
+				// so we have no way of knowing/checking ActiveInstanceID. To protect the endpoint
+				// from false DEL, we return the error without deleting/cleaning up.
+				return err
+			}
 		}
-		logger.WithField("stdin", args.StdinData).Debug("Updated stdin data for Delete Cmd")
-	}
 
-	ipamErr := ipam.ExecDel(conf.IPAM.Type, args.StdinData)
+		// Store WEP Metadata in endpoint only if it is not nil to avoid nil pointer dereference
+		// in some of the following operations when WEP doesn't exist in the datastore.
+		if wep != nil {
+			endpoint = wep.Metadata
+		} else {
+			endpoint = ep
+		}
 
-	if ipamErr != nil {
-		logger.Error(ipamErr)
+		// Check if ActiveInstanceID is populated (it will be an empty string "" if it was populated
+		// before this field was added to the API), and if it is there then compare it with ContainerID
+		// passed by the orchestrator to make sure they are the same, return without deleting if they aren't.
+		if endpoint.ActiveInstanceID != "" && args.ContainerID != endpoint.ActiveInstanceID {
+			logger.WithField("WorkloadEndpoint", wep).Warning("CNI_ContainerID does not match WorkloadEndpoint ActiveInstanceID so ignoring the DELETE cmd.")
+			return nil
+		}
 	}
 
-	calicoClient, err := CreateClient(conf)
-	if err != nil {
-		return err
+	// Store WEP Metadata in endpoint only if it is not nil to avoid nil pointer dereference
+	// in some of the following operations when WEP doesn't exist in the datastore.
+	if wep != nil {
+		endpoint = wep.Metadata
+	} else {
+		endpoint = ep
 	}
 
-	ep := api.WorkloadEndpointMetadata{
-		Name:         args.IfName,
-		Node:         nodename,
-		Orchestrator: orchestrator,
-		Workload:     workload}
-	if err = calicoClient.WorkloadEndpoints().Delete(ep); err != nil {
-		if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
-			logger.WithField("endpoint", ep).Info("Endpoint object does not exist, no need to clean up.")
-		} else {
+	// Delete the WorkloadEndpoint object from the datastore.
+	if err = calicoClient.WorkloadEndpoints().Delete(endpoint); err != nil {
+		switch err := err.(type) {
+		case errors.ErrorResourceDoesNotExist:
+			// Log and proceed with the clean up if WEP doesn't exist.
+			logger.WithField("endpoint", endpoint).Info("Endpoint object does not exist, no need to clean up.")
+		case errors.ErrorResourceUpdateConflict:
+			// This case means the WEP object was modified between the time we did the Get and now,
```

Comment:
I don't think retrying makes sense from CNI side since if some data changed while we were trying to Compare-and-Delete then retrying with the same data will still fail. 
It is the orchestrator's (k8s) job to send another DEL with the correct/different ContainerID if CNI plugin returns an error when k8s sent the DEL. k8s will do the retry here